runtime/commonMain/src/kotlinx/serialization/internal/PrimitiveArraysSerializers.kt

```diff
@@ -0,0 +1,398 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.*
+
+private const val INITIAL_SIZE = 0
+
+// it could be an object, but https://youtrack.jetbrains.com/issue/KT-14628
+/**
+ * Serializer for [ByteArray].
+ *
+ * Encode elements one-by-one, as regular list,
+ * unless format's Encoder/Decoder have special handling for this serializer.
+ */
+public class ByteArraySerializer :
+    PrimitiveArraySerializer<Byte, ByteArray, ByteArraySerializer.Builder>(ByteSerializer, ByteDescriptor),
+    KSerializer<ByteArray> {
+    inner class Builder internal constructor() :
+        PrimitiveArraySerializer<Byte, ByteArray, ByteArraySerializer.Builder>.Builder() {
+
+        private var buf: ByteArray = ByteArray(INITIAL_SIZE)
```

Comment:
Performance-wise observation:

1) Most of the arrays definitely have the size bigger than 1, first K resizes will be wasteful. Let's stick to the Java and make it 10 by default.
2) 0 in size requires you to resize as `* 2 + 1`  instead of plain `* 2`
3) Resize is currently quadratic because on each append you ensure that size is at least "position * 2 + 1", effectively doing resize on **each** append